apps/src/lab2/views/components/Instructions.tsx

```diff
@@ -134,22 +148,44 @@ const InstructionsPanel: React.FunctionComponent<InstructionsPanelProps> = ({
   message,
   messageIndex,
   imageUrl,
-  showNextButton,
+  showContinueButton,
+  showFinishButton,
+  beforeFinish,
   onNextPanel,
   layout = 'vertical',
   imagePopOutDirection = 'right',
   theme = 'dark',
   handleInstructionsTextClick,
 }) => {
   const [showBigImage, setShowBigImage] = useState(false);
+  const [isFinished, setIsFinished] = useState(false);
 
   const imageClicked = () => {
     setShowBigImage(!showBigImage);
   };
 
   const vertical = layout === 'vertical';
 
-  const canShowNextButton = showNextButton && onNextPanel;
+  const canShowContinueButton = showContinueButton && onNextPanel;
+
+  const canShowFinishButton = showFinishButton;
+
+  const currentLevelId = useAppSelector(state => state.progress.currentLevelId);
+
+  const onFinish = useCallback(() => {
+    if (beforeFinish) {
+      beforeFinish();
+    }
+    setIsFinished(true);
+  }, [beforeFinish]);
+
+  // When the level changes, the Finish button is active again if user returns to last level.
+  useEffect(() => {
```

Comment:
@sanchitmalhotra126 @cnbrenci - since you reviewed, I added this so that 'Finish' button is no longer disabled when user leaves the last level.

Reply:
Not blocking but does it work if you use `canShowFinishButton` as the dependency for this useEffect? i.e. reset the finish button state whenever we go from showing the finish button to not showing the finish button? Asking only because I'd like to avoid using redux in this inner component if possible (the idea was this inner component could be used standalone outside of the lab2/progress redux framework, but currently it's only used by the parent Instructions component, so not a big deal).

Reply:
Oh no! I didn't see this until after I merged. Sorry about that! I can do a fast folllow-up for this. Thanks!

Reply:
https://github.com/code-dot-org/code-dot-org/pull/58528